throughput/spam.go

```diff
@@ -0,0 +1,427 @@
+	sZipf    float64
+	vZipf    float64
+
+	// TPS parameters
+	txsPerSecond     int
+	minTxsPerSecond  int
+	txsPerSecondStep int
+	numClients       int // Number of clients per uri node
+
+	// Number of accounts
+	numAccounts int
+}
+
+func NewSpammer(
+	uris []string,
+	key *auth.PrivateKey,
+	balance uint64,
+	sZipf, vZipf float64,
+	txsPerSecond, minTxsPerSecond, txsPerSecondStep, numClients, numAccounts int,
+) *Spammer {
+	// Log Zipf participants
+	zipfSeed := rand.New(rand.NewSource(0))
+
+	return &Spammer{
+		uris,
+		key,
+		balance,
+		zipfSeed,
+		sZipf,
+		vZipf,
+		txsPerSecond,
+		minTxsPerSecond,
+		txsPerSecondStep,
+		numClients,
+		numAccounts,
+	}
+}
+
+// Spam tests the throughput of the network by sending transactions using
+// multiple accounts and clients. It first distributes funds to the accounts
+// and then sends transactions between the accounts. It returns the funds to
+// the original account after the test is complete.
+// [sh] injects the necessary functions to interact with the network.
+// [terminate] if true, the spammer will stop after reaching the target TPS.
+// [symbol] and [decimals] are used to format the output.
+func (s *Spammer) Spam(ctx context.Context, sh SpamHelper, terminate bool, symbol string) error {
+	// log distribution
+	s.logZipf(s.zipfSeed)
+
+	// new JSONRPC client
+	cli := jsonrpc.NewJSONRPCClient(s.uris[0])
+
+	factory, err := auth.GetFactory(s.key)
+	if err != nil {
+		return err
+	}
+
+	// Compute max units
+	parser, err := sh.GetParser(ctx)
+	if err != nil {
+		return err
+	}
+	actions := sh.GetTransfer(s.key.Address, 0, uniqueBytes())
+	maxUnits, err := chain.EstimateUnits(parser.Rules(time.Now().UnixMilli()), actions, factory)
+	if err != nil {
+		return err
+	}
+
+	unitPrices, err := cli.UnitPrices(ctx, false)
+	if err != nil {
+		return err
+	}
+	feePerTx, err := fees.MulSum(unitPrices, maxUnits)
+	if err != nil {
+		return err
+	}
+
+	var fundsL sync.Mutex
+	// distribute funds
+	accounts, funds, factories, err := s.distributeFunds(ctx, cli, parser, feePerTx, sh)
+	if err != nil {
+		return err
+	}
+
+	// create issuers
+	issuers, err := s.createIssuers(parser)
+	if err != nil {
+		return err
+	}
+
+	// make sure we can exit gracefully & return funds
+	signals := make(chan os.Signal, 2)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
+	cctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, issuer := range issuers {
+		issuer.Start(cctx)
+	}
+
+	// set logging
+	t := issuers[0].logStats(cctx)
+	defer t.Stop()
+
+	// Broadcast txs
+	var (
+		// Do not call this function concurrently (math.Rand is not safe for concurrent use)
+		z = rand.NewZipf(s.zipfSeed, s.sZipf, s.vZipf, uint64(s.numAccounts)-1)
+
+		it                      = time.NewTimer(0)
+		currentTarget           = min(s.txsPerSecond, s.minTxsPerSecond)
+		consecutiveUnderBacklog int
+		consecutiveAboveBacklog int
+
+		stop bool
+	)
+	utils.Outf("{{cyan}}initial target tps:{{/}} %d\n", currentTarget)
```

Comment:
(followup) Could we break the section broadcasting txs into its own function as well? Separating the spammer set up into functions much easier to review